### ISSUE TITLE
Too much recursion because of @Injectable conflicts with other decorators

### ISSUE THREAD

**Issue by tamtakoe** [angular/angular#35733]
# 🐞 bug report

### Affected Package
Looks like `@angular/core`

### Is this a regression?
Yes, the versions before 9 work correct (https://stackblitz.com/edit/angular-3sutb9 - works, because stackblitz is not provided Ivy-rendering)

### Description
`@injectable` conflicts with other decorators in this example:
```
@Injectable()
export class TestServiceA {}

@Injectable()
@testDecorator()
export class TestServiceB extends TestServiceA {}

@Injectable()
export class TestService extends TestServiceB {}

@Component({
  selector: 'my-app',
  templateUrl: './app.component.html',
  providers: [TestService]
})
export class AppComponent  {
  constructor(private testService: TestService) {}
}
```

## 🔬 Minimal Reproduction
You can just clone the repo and run locally (stackblitz doesn't work with last versions of Angular)
https://github.com/tamtakoe/angular-max-call-stack-size

## 🔥 Exception or Error
<pre><code>core.js:3866 ERROR RangeError: Maximum call stack size exceeded
    at TestServiceB_Factory (app.component.ts:30) ...

or  ERROR InternalError: "too much recursion" ... in Firefox etc.
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>Angular CLI: 9.0.3
Node: 12.4.0
OS: darwin x64

Angular: 9.0.4
... common, compiler, compiler-cli, core, forms
... language-service, localize, platform-browser
... platform-browser-dynamic, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.3
@angular-devkit/build-angular     0.900.3
@angular-devkit/build-optimizer   0.900.3
@angular-devkit/build-webpack     0.900.3
@angular-devkit/core              9.0.3
@angular-devkit/schematics        9.0.3
@angular/cli                      9.0.3
@ngtools/webpack                  9.0.3
@schematics/angular               9.0.3
@schematics/update                0.900.3
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2
</code></pre>

**Comment by PapaNappa** [comment]
I have a similar error after upgrading to 9.0.4/0.900.4:

```
Compiling @angular/… : es2015 as esm2015

chunk {} runtime.689ba4fd6cadb82c1ac2.js (runtime) 1.45 kB [entry] [rendered]
chunk {1} main.9d65be46ac259c9df33b.js (main) 700 kB [initial] [rendered]
chunk {2} polyfills.da2c6c4849ef9aea9de5.js (polyfills) 35.6 kB [initial] [rendered]
chunk {3} styles.1b30f95df1bf800225e4.css (styles) 62.7 kB [initial] [rendered]
Date: 2020-02-28T12:40:47.476Z - Hash: 5d47462af19da00a8547 - Time: 94639ms

ERROR in Error when flattening the source-map "<path1>.d.ts.map" for "<path1>.d.ts": RangeError: Maximum call stack size exceeded
ERROR in Error when flattening the source-map "<path2>.d.ts.map" for "<path2>.d.ts": RangeError: Maximum call stack size exceeded
ERROR in Error when flattening the source-map "<path3>.d.ts.map" for "<path3>.d.ts": RangeError: Maximum call stack size exceeded
ERROR in Error when flattening the source-map "<path4>.d.ts.map" for "<path4>.d.ts": RangeError: Maximum call stack size exceeded
…
```

Unfortunately, I was not yet able to reproduce this issue.
This appears to happen only once after upgrading an existing Angular <9.0.4/3 (not sure which one) project to 9.0.3/4. I can run `ng build` again and then it works flawlessly. Even with `npm ci` I cannot reproduce the error. :/

**Comment by pkozlowski-opensource** [comment]
This seems to be a legit report of the behaviour change in ivy with custom decorators (there were multiple similar bugs reported previously), here is a live repro: https://ng-run.com/edit/IneGzHSECoI3LtdEXPWh

**Comment by JoostK** [comment]
There's several issues here.

First, this line introduces the infinite loop:

```ts
newConstructor.prototype = Object.create(target.prototype);
```

This effectively sets up `target` to act as a superclass of `newConstructor`, as far as I understand it. Then, when the runtime looks up the factory for `TestServiceB` it delegates to the parent (as there's no own constructor) here:

https://github.com/angular/angular/blob/d2c60cc216982ee018e7033f7f62f2b423f2d289/packages/core/src/render3/di.ts#L660

Here, `type.prototype === newConstructor.prototype`, so using `Object.getPrototypeOf` we arrive at `target.prototype` which is `TestServiceB`. Therefore, the superclass' factory that is obtained for `TestServiceB` *is* `TestServiceB` again, causing the infinite loop.

**Disclaimer**: I am always confused by prototype stuff in JS, so there could be nonsense in the above.

---

Secondly, the compiled definitions that are present as static property on the class are not copied over to `newConstructor`. This is a problem for DI to work correctly, as factory functions etc will not be available. You don't currently see the effect of this as none of the constructors have parameters that need to be injected.

**Comment by splincode** [comment]
@atscott Is this added backport to Angular 9?

I see only add to Angular 10

![image](https://user-images.githubusercontent.com/12021443/84540638-8322fb80-acfe-11ea-9aa1-d0b76eac6a6c.png)

**Comment by atscott** [comment]
@splincode This change wasn't merged into the 9.1.x branch. At the time the fix was created, `master and patch` branches would have referred to `10.0 and 9.1` respectively. As of today (and 9 days ago when it was merged), the two branches were `10.1 and 10.0`. Thanks for bringing this up. @crisbeto - Should we create a new PR to target `9.1`?

**Comment by atscott** [comment]
@splincode This will be available in the `9.1.12` release next week.

**Comment by splincode** [comment]
@atscott 11 days ago `9.1.12` not found))
